### ISSUE TITLE
Insert statements are painfully slow. When stopped, application crashes

### ISSUE THREAD

**Issue by TheSynthMaster** [sqlitebrowser/sqlitebrowser#3237]
I was inserting a large number of records via insert statement (importing data from an Oracle DB)

While each sql statement execute in 50-55 ms, only one or two each second are processed, so totally useless for 100k rows.
Moreover, when pressing the stop button, app crashes, the DB is not modified.

Useful extra information
-------------------------
> DB4S v3.12.2 [built for x86_64-little_endian-llp64] on Windows 10 (10.0) (winnt/10.0.19045) [x86_64]
> using SQLite Version 3.35.5
> and Qt 5.12.8

**Comment by TheSynthMaster** [comment]
No, I did not tried... but how it's possible that the statements are executed so slowly? Each one is executed fast, it seems the parsing of the next statement who takes all the time... And the crashing bug when stopping the script is particularly nasty...

**Comment by chrisjlocke** [comment]
> And the crashing bug when stopping the script is particularly nasty

Have you tried the nightly at all?
https://nightlies.sqlitebrowser.org/latest/DB.Browser.for.SQLite-win32.zip

Extract the zip and run the .exe - nothing to install and it won't affect your currently installed version.
Both or none of these issues might be resolved/improved.

**Comment by MKleusberg** [comment]
Unfortunately the Execute SQL part was never really designed to work well with very large files. I guess the best approach here is to open your database file, then click File -> Import -> Database from SQL file and select your Oracle export. DB4S will then ask you whether to import into the currently opened database. Click Yes there and the import will start and run faster. That said, I'll also see if I can speed up the Execute SQL area a bit.

**Comment by MKleusberg** [comment]
I just made some changes to speed this up. You can try again using tomorrow's nightly build. For what it's worth, in my test case using ~10,000 INSERT statements from a ~1.2MiB file the speed up is at around 5x. The larger the file though the more significant the possible speed up. So in your case it might be different.

There's another opportunity for optimisation here. Most of the remaining wasted time is spend on logging the executed statements. Disabling that gives 5x speed up in my test case. On the other hand the SQL log is obviously an important feature and we shouldn't just get rid of it. One idea would be to limit it to a max size and start removing rows in the front. I'm not sure though whether the time we can save by doing this is greater than the time required for checking and manipulating the contents. Also it's not obvious to me what the maximum number of lines should be (yes, another option in the settings dialog would be one way to solve this but we already have so many of those and we should still set a sensible default then). @sqlitebrowser/developers Any ideas? Or should we leave it and consider this good enough?

**Comment by justinclift** [comment]
Some kind of `[ ] Disable logging` toggle option somewhere is also my first thought, but as you point out... with a bit more thinking there might be a better approach. :smile:

**Comment by mgrojo** [comment]
I haven't looked at it, but I think we are doing something redundant when running more than one statement. We sent (more or less) the same output to the "SQL Log" and to the bottom results pane. In the "SQL Log" it makes sense, because it is additive, but in the results pane, it is overwritten continually. Maybe we should stop overwriting on each statement execution and only write it at the end? When running one statement, both outputs should still show the same result, but when running more than one, we could write in the results pane the global statistics, i.e. how much time everything took, how many statements were executed, and maybe direct the user to  "SQL Log > User" for the complete details. We could even add more useful global statistics, like the statement that peaked in time, or how many rows were affected in total.

**Comment by mgrojo** [comment]
Don't know if that would speed up anything, it actually depends on how costly it's writing in the results pane.

**Comment by sky5walk** [comment]
I am surprised the log was this heavy? I would prefer the option to disable logging except for errors.

If that truly speeds insertions.
Are there log delays on copying many rows into clipboard? That is also slow.

**Comment by MKleusberg** [comment]
I've also tried disabling the bottom results pane but it didn't really have any measurable effect. A lot of time for the SQL logging is spent in Scintilla (that's the text editor widget we use) routines. Syntax highlighting and code folding take some time but most time really seems to be wasted on text insertion itself. This made me think of a new approach that I've just implemented: assembling the log text silently while the execution runs and only making one single large update of the GUI after the last SQL statement has finished. This gives us almost the full speed up (one minor delay can be noticed when updating the GUI) without losing much functionality.
